man/cpddumbfs.8
.TH "CPDDUMBFS" "8" "January 11, 2013" "1.0" "ddumbfs"
.SH NAME
cpddumbfs \- command line tools to copy from and to a offline ddumbfs filesystem
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
\fBcpddumbfs\fP [options] \fIsource\fP \fItarget\fP
.SH DESCRIPTION
.sp
\fBcpddumbfs\fP copy from and to an offline ddumbfs filesystem
.sp
One and only one of the \fIsource\fP or \fItarget\fP must be inside the \fIddfsroot\fP directory.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-h, \-\-help
show this help message and exit
.UNINDENT
.INDENT 0.0
.TP
.B \-v, \-\-verbose
display addresses and hashes from file (download only)
.UNINDENT
.INDENT 0.0
.TP
.B \-l, \-\-lock_index
lock index into memory (increase speed for large file)
.UNINDENT
.INDENT 0.0
.TP
.B \-c, \-\-check
check file integrity in exit code (download only)
.UNINDENT
.INDENT 0.0
.TP
.B \-f, \-\-force
download a file even if the header is corrupted or if \fIbad blocks\fP
are found.
.UNINDENT
.SH EXAMPLE
.sp
usage:
.sp
.nf
.ft C
$ cpddumbfs \-l /data/ddumbfs/ddfsroot/data /tmp
$ cpddumbfs /tmp/data /data/ddumbfs/ddfsroot
.ft P
.fi
.SH REMARKS
.sp
\fIcpddumbfs\fP is single threaded. If you have a multi\-core CPU, \fIcpddumbfs\fP
will be slower than copying files to the online filesystem. By default
\fIddumbfs\fP mount the filesystem with a pool of thread used to calculate
block hash in parallel.
When downloading file from the offline filesystem, \fIcpddumbfs\fP is
as fast as the mounted filesystem.
.SH SEE ALSO
.sp
\fIddumbfs(1)\fP, \fImkddumbfs(8)\fP
.SH AUTHOR
.sp
Alain Spineux <\fI\%alain.spineux@gmail.com\fP>
.SH AUTHOR
Alain Spineux
.SH COPYRIGHT
2011, Alain Spineux
.\" Generated by docutils manpage writer.
.\" 
.


doc/Changelog


2013-01-12 ddumbfs-1.1

    fsckddumbfs:
      - option pack to pack the block file and be able to downsize the filesystem using migrateddumbfs

    all:
        - added .deb generation
        - improved .rpm generation

    ddumbfs:
        - direct_io default is to be enable when block file is a block device
        - no more align option
        
    ddfslib:
        - bug: ddfs_init() was not repairing the blockfile MAGIC

    bits:
     - optimized and added more testing for quick_functions
        

2012-11-23 ddumbfs-1.0b21

    all:
      replaced "struct ddfs_ctx ddfs" by "struct ddfs_ctx *ddfs" to be able to 
      mount multiple filesystem, required by new migrateddumbfs 
      
    migrateddumbfs:
      - new tools to resize ddumbfs (migrate and existing fs into an fs of different size)
        
    ddumbfs:
      - don't start if .autofsck cannot be locked.

    fsckddumbfs:
      - bug: fsckddumbfs was unable to recreate an index.


2012-11-21 ddumbfs-1.0b20

    all:
      - changed logging to append mode to avoid a rollout every time ddumbfs or 
      any ddfs utils is started and finnaly loose log.  
      - don't check for libaio anymore in autoconf
      
    ddumbfs:
      - added ddumb_statistic.inode_counter and display ddumb_statistic.fh_counter

2012-11-17 ddumbfs-1.0b19
    
    bits:
      - added bit_array_cmp_count() 

    ddumbfs:
      - use bit_array_cmp_count() to reduce complexity 

    fsckddumbfs:
      - all block reading and hashing that were using libaio are no using 
      multi-threading to improve performance.  

2012-11-07 ddumbfs-1.0b18
    
    ddumbfs:
      - bug: ddfs_save_usedblocks() was not SYNCING bfile and ifile, now it does
      - added c_auto_buffer_flush=60s to force flush of pending buffer and 
        c_auto_sync=120s to force sync of bfile and ifile. Can be tuned in ddfs.cfg
      
    fsckddumbfs:
      - bug fix that generate SIGSEV
      

2012-11-05 ddumbfs-1.0b17

    ddumbfs:
      - now ddumb_fsync() sync the bfile and the ifile too

    queryddumbfs:
      - added option "--force" to run even when filesystem is already mounted
      - it was not checking reslut of ddfs_read_full_block(), now it does


2012-10-31 ddumbfs-1.0b16

    ddumbfs:
      - lot of cosmetic change and renamed some variables and function 
      - fixed a bug in block_wait() 

2012-10-29 ddumbfs-1.0b15

    all:
      - close and closer to the 1.0 :-)

    ddumbfs:
      - bug: Access to file in .ddumbfs was locking the thread because these
      file dont have fh->lock and ddumbfs_read() was loccking fh->lockfor any
      file


2012-10-26 ddumbfs-1.0b14

    all:
      - in Makefile.am, replaced AM_LDFLAGS into LDADD, because of ubuntu 12.04

    ddumbfs:
      - flush "long living" buffer at regular interval 
      
    ddfschkrep:
      - check only regular files

    fsckddumbfs:
      - check only regular files


2011-11-21 ddumbfs-1.0b13

    all:
      - use bits mask and bits shift instead of mod, div and mul to get gap,
      and block_boundary from an offset.
    
    ddfslib      
      - improved ddfs_convert_addr() and ddfs_get_node_addr()

    ddumbfs:
      - the lock_index was not working because mlock() was done before
      the fork() done in fuse_main(). Now mlock is done in ddumbfs_init() 

2011-11-12 ddumbfs-1.0b12

    ddumbfs:
      - handle the unlikely race condition on block write using a list of block 
      being written. Any read check this list and must wait if the block is in the list
      - delayed error was wrongly reported 
      
    testddumbfs:
      - added operation 'B' to test for the 'race condition' above but this is 
      useless because FUSE cache the data and never issue a read.
      
      
2011-11-00 ddumbfs-1.0b11
    ddumbfs:
      - write requests are queued in a list and writers pick them up.
      - some tuning when waiting for a "REANDONLY" fh to be written (in _ddumb_write()
      and do_truncate())

    !! released soon, mostly an alpha  

2011-10-25 ddumbfs-1.0b10

    mkddumbfs:
      - bug: 128K block is the default
      - bug: c_addr_size was wrongly calculated, 1 bit too small and then
      because of the use of octet, 1/8 wrong !
      
    queryddumbfs:
      - NEW tool to query and search the ddumbfs index and files about nodes, 
      hash and blocks.

2011-10-25 ddumbfs-1.0b9
    xlog:
      - was rolling out twice at startup when not in "append" mode

    ddumbfs:
      - removed xstat->ref because was xstat->fhs_n similar 
      - bug: in writer_pool_handle() "fh_dst->pool_loaded=1" moved before xstat_subscribe() 
      - added option "reclaim" to handle auto reclaim when disk usage > reclaim%
       

2011-10-23 ddumbfs-1.0b8
    xlog:
      - added filename alias, to handles file like 'stderr' 
      - error handling
    ddumbfs:
      - close file after read: "/proc/cpuinfo"
      - reclaim() handle non regular file like (sockets, fifos, block
      & char devices) better
      - symblink set the uid and gid appropriately
      - pjd succeed 
      - added option [no]fuse_default that insert: "-ouse_ino,readdir_ino,
         ...default_permissions,allow_other,big_writes,max_read=131072,
         ...max_write=131072,negative_timeout=0,entry_timeout=0,attr_timeout=0
      - "fuse_default" can be over-written
      - bug in xstat_compare() was generating "xstat_release: xstat not found, continue"
      use long long int insead of int
      - complete rewrite of the writer pool. Writers are in a list, the requester
      iterates the list until a free writer is found, upload data in the ddumb_fh
      and signal the thread.

    ddfslib:
      - close file after read: "ddfs.cfg"
    mkddumbfs:
      - default block size is now 128k
    cpddumbfs:
      - added some code to "download" file even if bad block are found.
      
2011-10-10 ddumbfs-1.0b7
    ddumbfs:
      - added late block allocation to reduce self-fragmentation and increase

src/cpddumbfs.c
/**
 * @cpddumbfs.c
 * @author  Alain Spineux <alain.spineux@gmail.com>
 * @version 0.0
 *
 * @section LICENSE
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details at
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @section DESCRIPTION
 *
 * cpddumbfs copy from and to a offline ddumbfs filesystem
 *
 */

#define _XOPEN_SOURCE 500
#define _LARGEFILE64_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <ftw.h>
#include <fcntl.h>
#include <assert.h>

#include "ddfslib.h"

#define MAX_SIZE 0xFFFFFFFFFFFFFFFFll;

int verbose_flag=0;
int check_integrity=0;
int lock_index_flag=0;
int force_flag=0;

static struct option long_options[] =
{
       /* These options set a flag. */
       {"help",          no_argument,       0, 'h'},
       {"lock_index",    no_argument,       0, 'l'},
       {"check",         no_argument,       0, 'c'},
       {"verbose",       no_argument,       0, 'v'},
       {"force",         no_argument,       0, 'f'},
       {0, 0, 0, 0}
};

void usage()
{
    printf("Usage: cpddumbfs [options] source target\n"
            "\n    copy from and to an offline ddumbfs filesystem\n\nOptions\n"
            "  -h, --help            show this help message and exit\n"
            "  -v, --verbose         display addresses and hashes from file (download only)\n"
            "  -l, --lock_index      lock index into memory (increase speed for large file)\n"
            "  -c, --check           check file integrity in exit code (download only)\n"
            "  -f, --force           download a file even with a corrupted MAGIC in the header\n"
            "\n  One and only one of the source or target must be inside the\n"
            "  ddfsroot directory. Use - to redirect from/to stdin/stdout.\n"
            "\nSamples:\n"
            "  cpddumbfs -l /data/ddumbfs/ddfsroot/data /tmp\n"
            "  cpddumbfs /tmp/data /data/ddumbfs/ddfsroot\n"
    );
}

int download(char *source, char *destination)
{
    char dstfilename[FILENAME_MAX];
    unsigned char node[NODE_SIZE];
    uint64_t size;
    int len;

    int integrity=1;

    if (verbose_flag) fprintf(stderr, "download %s %s\n", source, destination);
    int fsrc=open(source, O_RDONLY);
    if (fsrc==-1)
    {
        perror(source);
        return 1;
    }

    int fdst=-1;
    if (0==strcmp("-", destination))
    {
        fdst=1; // stdout
    }
    else
    {
        if (isdir(destination))
        {
            char *p=strrchr(source, '/');
            if (p==NULL) p=source;
            else p=p+1;
            snprintf(dstfilename, FILENAME_MAX, "%s/%s", destination, p);
        }
        else strncpy(dstfilename, destination, FILENAME_MAX);

        fdst=open(dstfilename, O_RDWR | O_CREAT | O_TRUNC, 0600);
        if (fdst==-1)
        {
            perror(dstfilename);
            goto ERROR1;
        }
    }

    // read file size
    len=file_header_set_conv(fsrc, &size);

    if (len==-1)
    {
        perror(source);
        goto ERROR1;
    }
    else if (len==0)
    {
        fprintf(stderr, "magic header missing, not a valid ddumbfs file: %s\n", source);
        if (!force_flag) goto ERROR1;
    }
    else if (len!=ddfs->c_file_header_size)
    {
        fprintf(stderr, "cannot read file header size: %s\n", source);
        if (!force_flag) goto ERROR1;
    }

    long long int file_size=size;
    if (len!=ddfs->c_file_header_size)
    {
        fprintf(stderr, "unknown file size, continue: %s\n", source);
        file_size=-1;
        integrity=0;
    }

    long long int i=0;
    long long int write_size=0;

    int nlen=ddfs->c_node_size;
    while (1 || (force_flag && file_size==-1)) // || write_size<file_size)
    {
        int fake_node=0;
        int fake_block=0;
        long long int fileoff=ddfs->c_file_header_size+i*ddfs->c_node_size;
        nlen=pread(fsrc, node, ddfs->c_node_size, fileoff);
        if (nlen==0) break; // EOF
        if (nlen!=ddfs->c_node_size)
        {
            if (nlen==-1) fprintf(stderr, "error reading offset %lld (%s)\n", fileoff, strerror(errno));
            else fprintf(stderr, "error reading offset %lld short read %d/%d\n", fileoff, nlen, ddfs->c_node_size);
            if (!force_flag) goto ERROR1;
            memset(node, '\0', ddfs->c_node_size);
            ddfs_convert_addr(1, node);
            integrity=0;
            fake_node=1;
        }

        blockaddr addr=ddfs_get_node_addr(node);
        len=ddfs_read_full_block(addr, ddfs->aux_buffer);
        if (len!=ddfs->c_block_size)
        {
            if (len==-1) fprintf(stderr, "error reading block file offset %lld (%s)\n", addr*ddfs->c_block_size, strerror(errno));
            else fprintf(stderr, "error reading block file offset %lld short read %d/%d\n", addr*ddfs->c_block_size, len, ddfs->c_block_size);
            integrity=0;
            if (!force_flag) goto ERROR1;
            ddfs_forced_read_full_block(addr, ddfs->aux_buffer, 1024);
            len=ddfs->c_block_size;
            fake_block=1;
        }

        char *status="";
        int badblock=0;
        if (addr==1)
        {
            badblock=1;
            status="corrupted";
        }

        if (check_integrity && !fake_node)
        {
            status="ok";
            if (addr==0)
            {   // check hash match zero block
                if (0!=memcmp(ddfs->null_block_hash, node+ddfs->c_addr_size, ddfs->c_hash_size) && 0!=memcmp(ddfs->zero_block_hash, node+ddfs->c_addr_size, ddfs->c_hash_size))
                {
                    badblock=1;
                    status="bad"; // bad node
                }
            }
            else
            {
                unsigned char hash[HASH_SIZE];
                ddfs_hash(ddfs->aux_buffer, hash);
                if (memcmp(node+ddfs->c_addr_size, hash, ddfs->c_hash_size)!=0)
                {
                    badblock=1;
                    status="err";
                    // fprintf(stderr, "err block %6lld %6lld %016llx<>%016llx\n", i, addr, *(long long int*)(node+ddfs->c_addr_size), *(long long int*)hash);
                    // search the Index, maybe I can found the good hash ?
                    blockaddr baddr;
                    nodeidx node_idx=ddfs_search_hash(node+ddfs->c_addr_size, &baddr);
                    //fprintf(stderr, "err node_idx=%lld baddr=%lld\n", node_idx, baddr);
                    while (badblock && 0<=node_idx && node_idx<ddfs->c_node_count && 0==memcmp(node+ddfs->c_addr_size, ddfs->nodes+(ddfs->c_node_size*node_idx+ddfs->c_addr_size), ddfs->c_hash_size))
                    {
                        baddr=ddfs_get_node_addr(ddfs->nodes+(ddfs->c_node_size*node_idx));
                        // fprintf(stderr, "test node node_idx=%lld baddr=%lld\n", node_idx, baddr);
                        if (baddr!=addr && baddr!=1)
                        {
                            len=ddfs_read_full_block(baddr, ddfs->aux_buffer);
                            if (len!=ddfs->c_block_size)
                            {
                                if (len==-1) fprintf(stderr, "error reading block file offset %lld (%s)\n", baddr*ddfs->c_block_size, strerror(errno));
                                else fprintf(stderr, "error reading block file offset %lld short read %d/%d\n", baddr*ddfs->c_block_size, len, ddfs->c_block_size);
                                if (!force_flag) goto ERROR1;
                                ddfs_forced_read_full_block(baddr, ddfs->aux_buffer, 1024);
                                integrity=0;
                                len=ddfs->c_block_size;
                                fake_block=1;
                            }
                            ddfs_hash(ddfs->aux_buffer, hash);
                            if (0==memcmp(node+ddfs->c_addr_size, hash, ddfs->c_hash_size))
                            {
                                badblock=0;
                                status="corrected";
                                addr=baddr;
                            }
                        }
                        node_idx++;
                    }
                }

            }
        }

        if (badblock) integrity=0;

        long long int sz=ddfs->c_block_size;
        if (file_size!=-1)
        {
            sz=file_size-write_size;
        }
        if (sz>ddfs->c_block_size) sz=ddfs->c_block_size;

        len=write(fdst, ddfs->aux_buffer, sz);
        if (len!=sz)
        {
            if (len==-1) fprintf(stderr, "error writing file offset %lld (%s)\n", write_size, strerror(errno));
            else fprintf(stderr, "error writing file, short write offset %lld %d/%lld\n", write_size, len, sz);
            goto ERROR1; // no force_flag here
        }
        if (verbose_flag || 0!=strcmp(status, "ok"))
        {
            fprintf(stderr, "%6lld %6lld %016llx %s\n", i, addr, (long long int)ddfs_hton64(*(uint64_t*)(node+ddfs->c_addr_size)), status);
        }

        write_size+=sz;
        i++;
    }

    if (verbose_flag)
    {
        if (file_size>0) fprintf(stderr, "        size: %lld\n", file_size);
        else fprintf(stderr, "        size: NA\n");
        fprintf(stderr, "written size: %lld\n", write_size);
    }

    close(fsrc);
    if (fdst!=1) close(fdst);

    return !integrity;

    ERROR1:
    if (fdst!=1 && fdst!=-1) close(fdst);
    close(fsrc);
    return 1;
}


int upload(char *source, char *destination)
{
    char dstfilename[FILENAME_MAX];
    uint64_t size;
    int len, res;

    if (ddfs->auto_fsck || ddfs->rebuild_fsck)
    {
        fprintf(stderr, "The filesystem must be checked.\n");
        return 1;
    }

    if (verbose_flag) fprintf(stderr, "upload %s %s\n", source, destination);
    int fsrc;
    if (0==strcmp("-", source))
    {
        fsrc=0; // stdin
    }
    else
    {
        fsrc=open(source, O_RDONLY);
        if (fsrc==-1)
        {
            perror(source);
            return 1;
        }
    }

    int fdst;
    if (isdir(destination))
    {
        if (fsrc==1)
        {
            fprintf(stderr, "destination must have a name when source is stdin\n");
            return 1;
        }
        char *p=strrchr(source, '/');
        if (p==NULL) p=source;
        else p=p+1;
        snprintf(dstfilename, FILENAME_MAX, "%s/%s", destination, p);
    }
    else strncpy(dstfilename, destination, FILENAME_MAX);

    fdst=open(dstfilename, O_RDWR | O_CREAT | O_TRUNC, 0600);
    if (fdst==-1)
    {
        perror(dstfilename);
        close(fsrc);
        return 1;
    }

    res=lseek(fdst, ddfs->c_file_header_size, SEEK_SET);
    if (res==-1)
    {
        perror(source);
        return 1;
    }

    size=0;
    len=ddfs->c_block_size;
    long long int i=0;
    while (len==ddfs->c_block_size)
    {
        unsigned char node[NODE_SIZE];
        long long int addr;

        len=read(fsrc, ddfs->aux_buffer, ddfs->c_block_size);
        if (len==-1)
        {
            perror(source);
            return 1;
        }
        size+=len;
        if (len<ddfs->c_block_size) memset(ddfs->aux_buffer+len, '\0', ddfs->c_block_size-len);
        addr=ddfs_write_block(ddfs->aux_buffer, node+ddfs->c_addr_size);
        ddfs_convert_addr(addr, node);

        if (verbose_flag)
        {
            fprintf(stderr, "%6lld %6lld %016llx...\n", i, addr, (long long int)ddfs_hton64(*(uint64_t*)(node+ddfs->c_addr_size)));
        }

        res=write(fdst, node, ddfs->c_node_size);
        if (res!=ddfs->c_node_size)
        {
            fprintf(stderr, "error writing file\n");
            return 1;
        }
        i++;
    }

    // write size
    len=file_header_get_conv(fdst, size);
    if (len==-1)
    {
        perror(dstfilename);
        return 1;
    }
    else if (len!=ddfs->c_file_header_size)
    {
        fprintf(stderr, "cannot write file header\n");
        return 1;
    }

    if (fsrc!=0) close(fsrc);
    close(fdst);

    return 0;

}

struct ddfs_ctx ddfsctx;

int main(int argc, char *argv[])
{
    int res=0;
    int c;

    ddfs=&ddfsctx;
    while (1)
    {
        // getopt_long stores the option index here.
        int option_index = 0;

        c=getopt_long(argc, argv, "hvlcf", long_options, &option_index);

        // Detect the end of the options.
        if (c==-1) break;

        switch (c)
        {

            case 'h':
                usage();
                return 0;
                break;

            case 'v':
                verbose_flag=1;
                break;

            case 'l':
                lock_index_flag=1;
                break;

            case 'c':
                check_integrity=1;
                break;

            case 'f':
                force_flag=1;
                break;

            case '?':
                // getopt_long already printed an error message.
                break;

            default:
                abort();
        }
    }

    if (optind+1>=argc)
    {
        fprintf(stderr, "missing source or target directory %d %d!\n", optind, argc );
        return 1;
    }

    char pdir[FILENAME_MAX];
    int up=0;
    int down=0;

    if (ddfs_find_parent(argv[optind], pdir)==0) down=1;
    if (ddfs_find_parent(argv[optind+1], pdir)==0) up=1;

    if (down && up)
    {
        fprintf(stderr, "Only one of the source or destination can be inside the offline ddumbfs filesystem.\n");
        return 1;
    }

    if (up && is_mounted(pdir))
    {   // yes you can download from a mounted filesystem :-)
        fprintf(stderr, "filesystem already mounted: %s\n", pdir);
        return 1;
    }

    if (ddfs_loadcfg(pdir, NULL))
    {
        return 1;
    }

    if (ddfs_init(DDFS_NOFORCE, DDFS_NOREBUILD, DDFS_DIRECTIOAUTO, lock_index_flag, NULL))
    {
        return 1;
    }

    if (down)
    {
        res=download(argv[optind], argv[optind+1]);
        if (check_integrity)
        {
            if (res) fprintf(stderr, "ERR\n");
            else fprintf(stderr, "OK\n");
        }
    }
    else if (up)
    {
        res=upload(argv[optind], argv[optind+1]);
    }

    ddfs_close();
    return res;

}
